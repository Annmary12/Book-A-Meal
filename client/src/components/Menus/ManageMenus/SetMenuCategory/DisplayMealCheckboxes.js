import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import styles from '../set-menu.scss';

/**
 * @class DisplayMealCheckboxes
 */
class DisplayMealCheckboxes extends Component {
  handleCheck = meal => () => {
    const { toggleMealInCategory } = this.props;
    toggleMealInCategory(meal);
  }

  /**
   * @returns {JSX} React JSX
   */
  render() {
    const {
      meals,
      categoryMealIds,
      errors,
      categoryIndex
    } = this.props;

    const error = errors[`categories[${categoryIndex}].meals`];

    return (
      <div>
        <b>Meals:</b>
        <div
          className={classnames({
            [styles.meals]: true,
            [styles.error]: error
          })}
        >
          {
            meals.map(meal => (
              <div key={meal.id}>
                <label className={styles.checkBoxLabel}>
                  {meal.title}
                  <input
                    type="checkbox"
                    checked={categoryMealIds.includes(meal.id) ? 'checked' : ''}
                    onChange={this.handleCheck(meal)}
                  />
                  <span className={styles.checkmark} />
                </label>
              </div>
            ))
          }
        </div>
        {
        error &&
        <div className={styles.mealsErrorMessage}>{error}</div>
      }
      </div>
    );
  }
}

DisplayMealCheckboxes.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  toggleMealInCategory: PropTypes.func.isRequired,
  categoryMealIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  errors: PropTypes.objectOf(PropTypes.string).isRequired,
  categoryIndex: PropTypes.number.isRequired
};

export default DisplayMealCheckboxes;
