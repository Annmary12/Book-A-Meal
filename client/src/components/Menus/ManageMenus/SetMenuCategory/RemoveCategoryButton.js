import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import cancelIcon from '../../../../../assets/img/cancel-icon.svg';
import styles from '../set-menu.scss';

const RemoveCategoryButton = ({
  handleRemoveCategory
}) => (
  <button
    className={classnames(styles.iconBtn, styles.right)}
    onClick={handleRemoveCategory}
  >
    <img src={cancelIcon} width="15" alt="cancel" />
  </button>
);

RemoveCategoryButton.propTypes = {
  handleRemoveCategory: PropTypes.func.isRequired
};

export default RemoveCategoryButton;
