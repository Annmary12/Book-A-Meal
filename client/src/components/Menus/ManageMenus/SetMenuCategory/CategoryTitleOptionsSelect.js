import React from 'react';
import PropTypes from 'prop-types';
import categoryTitles from '../categoryTitles.js';
import styles from '../set-menu.scss';
import RemoveCategoryButton from './RemoveCategoryButton.js';

const CategoryTitleOptionsSelect = ({
  selectedTitles,
  handleChange,
  removeCategory,
  categoryCount,
  value,
  errors,
  categoryIndex
}) => {
  const error = errors[`categories[${categoryIndex}].title`];

  return (
    <label htmlFor="title" className={styles.label}>
      <div className={styles.title}>
      Category title:
        {
        categoryCount > 1 &&
        <RemoveCategoryButton
          handleRemoveCategory={removeCategory}
        />
      }
      </div>
      <select
        className={error ? styles.inputError : styles.input}
        name="title"
        onChange={handleChange}
        value={value}
      >
        <option value="" disabled>Select a category</option>
        {
        categoryTitles.map((title, index) => (
          <option
            key={index.toString()}
            value={title}
            disabled={selectedTitles.includes(title) ? 'disabled' : ''}
          >
            {title}
          </option>
        ))
      }
      </select>
      {
        error &&
        <div className={styles.errorMessage}>{error}</div>
      }
    </label>
  );
};

CategoryTitleOptionsSelect.propTypes = {
  selectedTitles: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  removeCategory: PropTypes.func.isRequired,
  categoryCount: PropTypes.number.isRequired,
  errors: PropTypes.objectOf(PropTypes.string).isRequired,
  categoryIndex: PropTypes.number.isRequired
};
export default CategoryTitleOptionsSelect;
