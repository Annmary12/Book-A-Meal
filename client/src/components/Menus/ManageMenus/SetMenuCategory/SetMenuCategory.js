import React from 'react';
import PropTypes from 'prop-types';
import styles from '../set-menu.scss';
import CategoryTitleOptionsSelect from './CategoryTitleOptionsSelect';
import DisplayMealCheckboxes from './DisplayMealCheckboxes';

const SetMenuCategory = (props) => {
  const categoryMealIds = props.category.meals.map(meal => meal.id);
  return (
    <div className={styles.menuCategory}>
      <CategoryTitleOptionsSelect
        selectedTitles={props.selectedTitles}
        value={props.category.title}
        handleChange={props.handleCategoryTitleChange}
        removeCategory={props.removeCategory}
        categoryCount={props.categoryCount}
        errors={props.errors}
        categoryIndex={props.categoryIndex}
      />
      <DisplayMealCheckboxes
        meals={props.meals}
        categoryMealIds={categoryMealIds}
        toggleMealInCategory={props.toggleMealInCategory}
        errors={props.errors}
        categoryIndex={props.categoryIndex}
      />
    </div>
  );
};
SetMenuCategory.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  removeCategory: PropTypes.func.isRequired,
  categoryCount: PropTypes.number.isRequired,
  category: PropTypes.shape({
    meals: PropTypes.arrayOf(PropTypes.string),
    title: PropTypes.string
  }).isRequired,
  toggleMealInCategory: PropTypes.func.isRequired,
  handleCategoryTitleChange: PropTypes.func.isRequired,
  selectedTitles: PropTypes.arrayOf(PropTypes.string).isRequired,
  errors: PropTypes.objectOf(PropTypes.string).isRequired,
  categoryIndex: PropTypes.number.isRequired
};

export default SetMenuCategory;
