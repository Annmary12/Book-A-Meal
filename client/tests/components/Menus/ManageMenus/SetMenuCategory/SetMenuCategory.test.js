import React from 'react';
import SetMenuCategory from '../../../../../src/components/Menus/ManageMenus/SetMenuCategory/SetMenuCategory';

const props = {
  meals: [],
  removeCategory: jest.fn(),
  categoryCount: 1,
  category: {
    meals: [],
    title: ''
  },
  toggleMealInCategory: jest.fn(),
  handleCategoryTitleChange: jest.fn(),
  selectedTitles: [],
  errors: {},
  categoryIndex: 0
};

describe('SetMenuCategory components', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(
      <SetMenuCategory {...props}/>
    );

    expect(wrapper).toMatchSnapshot();
  })
});