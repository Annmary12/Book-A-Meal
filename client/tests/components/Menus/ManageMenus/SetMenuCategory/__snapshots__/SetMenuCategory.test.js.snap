// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetMenuCategory components should match snapshot 1`] = `
<div
  className="menuCategory"
>
  <CategoryTitleOptionsSelect
    categoryCount={1}
    categoryIndex={0}
    errors={Object {}}
    handleChange={[MockFunction]}
    removeCategory={[MockFunction]}
    selectedTitles={Array []}
    value=""
  />
  <DisplayMealCheckboxes
    categoryIndex={0}
    categoryMealIds={Array []}
    errors={Object {}}
    meals={Array []}
    toggleMealInCategory={[MockFunction]}
  />
</div>
`;
